avalon/pipeline.py

```diff
@@ -720,6 +720,84 @@ def get_representation_context(representation):
     return context
 
 
+def update_current_context(context):
+    """Update active Session to new context.
+
+    The context can hold `asset`, `task` and `app`, other keys are ignored.
+
+    Args:
+        context (dict): The context to apply.
+
+    Returns:
+        dict: The changed key, values in the current Session.
+
+    """
+
+    mapping = {
+        "AVALON_ASSET": context.get("asset", None),
+        "AVALON_TASK": context.get("task", None),
+        "AVALON_APP": context.get("app", None),
+    }
+    changed = {key: value for key, value in mapping.items() if
+               value and Session[key] != value}
+    if not changed:
+        return
+
+    # Update silo when asset changed
+    if "AVALON_ASSET" in changed:
+        asset_document = io.find_one({"name": changed['AVALON_ASSET'],
+                                      "type": "asset"},
+                                     projection={"silo": True})
+        assert asset_document, "Asset must exist"
+        changed["AVALON_SILO"] = asset_document['silo']
+
+    # Compute work directory (with the temporary changed session so far)
+    project = io.find_one({"type": "project"},
+                          projection={"config.template.work": True})
+    template = project['config']['template']['work']
+    _session = Session.copy()
+    _session.update(changed)
+    changed['AVALON_WORKDIR'] = format_template(template, _session)
+
+    # Update the full session in one go to avoid half updates
+    Session.update(changed)
+
+    # Emit session change
+    emit("currentContextUpdated", changed.copy())
```

Comment:
I would expect to get the Context from this signal, not the environment variables from a partial Session object. What do you think?

Reply:
This will pass on only whatever is changed. If at that point in time you'd like to get the active session state you can get it from `api.Session`.

Reply:
Ah wait, you meant it would pass on the `asset`, `task` and `app`  keys. Hmm. That could be interesting. How about passing both context *and* the changed environment?

Reply:
That's a good idea.

But first, there's one thing that's miscommunicated I think. `context` is a dictionary with a fixed membership, like specified in its schema. It should be passed whole to the function, and returned whole. No partials.

```python
def Context():
	return {
	    "project": api.Session["AVALON_PROJECT"],
	    "asset": api.Session["AVALON_ASSET"],
	    "silo": api.Session["AVALON_SILO"],
	    "subset": api.Session["AVALON_SUBSET"],
	    "version": api.Session["AVALON_VERSION"],
	    "representation": api.Session["AVALON_REPRESENTATION"]"
	}

context = Context()
context["project"] = "hulk"
context["asset"] = "Bruce"
api.update_current_context(context)
```

However now that I type that out, I realise not all members are in Session, and not all members can or should be changed this way, like Version and Representation..

That's a flaw in this design; we can't call it Context if it isn't going to be a Context else we'd confuse it with the use of Context elsewhere.. I'm not sure how to deal with this one.

My problem with passing `project=` and `asset=` directly is that it's not discoverable enough. Which members should be part of this function? Subset too? Silo? There's no way to tell without reading the docs or by trail and error. The thing about Context is that it's known and used in more than one place. It's worth learning about and doesn't come with surprises.

Any ideas?

Reply:
I'm inclined to go with..

```python
api.update_asset("Bruce")
api.update_task("modeling")
```

Or, though cringy..

```python
api.set_asset("Bruce")
```

Reply:
Ah, no. I see the confusion. The context with subset, version and representation is only relevant for a published file - exactly I'd go as far and let's say a *context* is always about a published file.

What we're updating here is the current Session, the current work area or workspace. It defines what "work environment" I'm currently in. That is **not** the context of a published file. This "work area" also has `app` and `task` which the context does not have - possibly even have `user` (I'd say).

I see to much issues with `update_asset` and `update_task` honestly. It might simplify some logic but it would ultimately also become slower.

> My problem with passing project= and asset= directly is that it's not discoverable enough. Which members should be part of this function? 

Hmm, interesting. I *love!* functions that expose the arguments with clear names, an IDE will let me know what I can pass along and often there's the docstring that complements the arguments.

I'd personally opt for either of:

1. Choose a single argument name that describes the dictionary in a clear concise way, e.g. `workspace` or `workarea` or `environment`
2. Give the function the exact arguments that are allowed, like `asset`, `task`, `app`. 

And additionally refactor the function name to something clearer, like `update_session_environment` or `update_session_workarea`?

Reply:
> I see to much issues with update_asset and update_task honestly. It might simplify some logic but it would ultimately also become slower and not necessarily safer because the whole work area is a "full set" working together.

Let's dive into that. I've got an idea of context switching will work from my days at Framestore and it's possible our ideas differ.

**Scenario 1**

1. User launches Maya, the current Session is fully qualified
2. User decides to switch from one asset to another
    - Clicks on the Current Asset icon, and is provided with a list of available assets
    - Picks one and Bam, asset updated.

**Scenario 2**

1. ...
2. User decides to switch project
    - Clicks on the Current Project icon, sees a list
    - Clicks on a relevant project
    - Current Asset button updates alongside it, to reflect that no asset has been set

In both of these cases, only one is ever updated at a time, and it's clear which of them does what.

With regards to performance.. I wouldn't worry about it. This could take seconds without being a problem.

> I love! functions that expose the arguments with clear names, an IDE will let me know what I can pass along and often there's the docstring that complements the arguments.

Agreed, but this isn't about arguments in general, but the fact that the function name doesn't hint towards what the arguments may be.

For example, consider this..

```python
api.update_task(...
```

What is the argument? Task!

```python
api.update_session(...
```

Now what is the argument? Task? Asset? Both? Either? A Session dictionary?

Your IDE may provide interactive hints, but a reader of the code and those without hints are left in the blind. Also I wouldn't lean on hints being available. They are an aid, not a necessity.

Reply:
Both your scenarios sound alright, albeit that *scenario 2* is not supported in Avalon because it would require the `io` to uninstall and reinstall to another project. As such the current updating does not allow to switch project.

Anyhow, I'm seeing this scenario most often:

Scenario 3
1. User decides to pick up another asset's task
  - Clicks on the asset
  - Clicks on the task
  - Sets the work area

They have now switched to asset X and task Y.

Reply:
I'm most concerned with "triggering updates" on the changes, say I'd like to initialize the new environment I have entered, but I don't now when I'm done switching to the new full context. Say I'm in task `modeling` for asset `hero` and want to go to asset `villain`'s task `lighting`.

If I do:
```python
api.set_asset("villain")
```

Will that now also update my `AVALON_WORKDIR` to `villain/modeling`? Would it even already start creating folders - say I'd want to automate it that far? How would I delay that until the work area switch was fully completed? (Say I'd want to avoid the `villain/modeling` to ever be initialized?

Reply:
Good point, how about we reset the task as well?

```
- project
   - asset
      - task
```

Then you could listen on just `taskChanged`.

Reply:
> Good point, how about we reset the task as well?

Yes, **but**... then your Scenario 1. My environment would not be initialized because I'm only listening to `taskChanged`.

Reply:
I'm not sure I see  the problem. Your workarea would have changed only when the task has changed, so changing the project wouldn't trigger any new folders to be created? Only when you change the task does that happen?

```python
api.set_asset("Bruce")
# No folders are created
api.set_task("animation")
# Bak, task folders created
```

What's the alternative?

Reply:
Hmm, actually. No? Unless when you do `api.set_asset("Bruce")` you are currently in limbo (no workspace) until you actually set a task.

Say I'm `Alice/modeling` and want to go to `Bruce/animation`.

```python
api.set_asset("Bruce")
# I'm now in `Bruce/modeling`; potentially I could start working! Actually, I just *entered* this workspace!
api.set_task("animation")
# I'm now in `Bruce/animation`
```

The alternatives are to "tell" when to enter the environment, where either:

1. You are in limbo until you trigger `api.enter_workarea()`
2. You have a single point of entry that capsulates your move to another work area, e.g. a single function.

Reply:
Ok, let's give your version a go. +1

Reply:
Thanks for the good discussion though, it will be important to remain critical to aspects like these. I'll update the PR and see where we get.

Reply:
No worries. I think that limbo is fine, but hard to say without actually using it.